src/boot/efi/boot.c

```diff
@@ -2268,6 +2306,104 @@ static VOID config_write_entries_to_variable(Config *config) {
+        static const UINT32 sb_vars_opts =
+                EFI_VARIABLE_NON_VOLATILE |
+                EFI_VARIABLE_BOOTSERVICE_ACCESS |
+                EFI_VARIABLE_RUNTIME_ACCESS |
+                EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS |
+                0;
+
+        static const struct {
+                const CHAR16 *name;
+                const CHAR16 *filename;
+                const EFI_GUID vendor;
+        } sb_vars[] = {
+                { L"db", L"db.auth", EFI_IMAGE_SECURITY_DATABASE_VARIABLE },
+                { L"KEK", L"KEK.auth", EFI_GLOBAL_VARIABLE },
+                { L"PK", L"PK.auth", EFI_GLOBAL_VARIABLE },
+        };
+
+        _cleanup_freepool_ CHAR8 *buffer = NULL;
+        UINTN size;
+        EFI_STATUS err = EFI_SUCCESS;
+
+        for (UINTN i = 0; i < ELEMENTSOF(sb_vars); i++) {
+                file_read(dir, sb_vars[i].filename, 0, 0, &buffer, &size);
+
+                if (!EFI_ERROR(err)) {
+                        err = efivar_set_raw(&sb_vars[i].vendor, sb_vars[i].name, buffer, size, sb_vars_opts);
+                        if (EFI_ERROR(err)) {
+                                Print(L"Failed to write %s secure boot variable: %r\n", sb_vars[i].name, err);
+                                uefi_call_wrapper(BS->Stall, 1, 3 * 1000 * 1000);
+                                return err;
+                        }
+                } else {
+                        Print(L"Failed reading %s file: %r\n", sb_vars[i].filename, err);
+                        uefi_call_wrapper(BS->Stall, 1, 3 * 1000 * 1000);
+                        return err;
+                }
+        }
+
+        Print(L"Finished loading secure boot variables.\n");
+        uefi_call_wrapper(BS->Stall, 1, 3 * 1000 * 1000);
+        return err;
+}
+
+static EFI_STATUS secure_boot_auto_enroll(Config *config, EFI_FILE *root_dir) {
+        EFI_STATUS err = EFI_SUCCESS;
+        EFI_FILE_HANDLE keys_dir;
+
+        if (setup_mode_enabled() && config->auto_enroll) {
+                Print(L"Both setup mode and auto enrollement are enabled.\n");
+
+                err = uefi_call_wrapper(root_dir->Open, 5, root_dir, &keys_dir, (CHAR16*) L"\\loader\\keys", EFI_FILE_MODE_READ, 0ULL);
+                if (!EFI_ERROR(err)) {
+                        for (;;) {
+                                CHAR16 buf[256];
+                                UINTN bufsize;
+                                EFI_FILE_INFO *f;
+
+                                bufsize = sizeof(buf);
+                                err = uefi_call_wrapper(root_dir->Read, 3, keys_dir, &bufsize, buf);
+                                if (bufsize == 0 || EFI_ERROR(err))
+                                        break;
+
+                                f = (EFI_FILE_INFO *) buf;
+                                if (f->FileName[0] == '.')
+                                        continue;
+
+                                if (f->Attribute & EFI_FILE_DIRECTORY) {
```

Comment:
invert condition please, and skip non-directories via `continue`. again, let's keep indetnation low